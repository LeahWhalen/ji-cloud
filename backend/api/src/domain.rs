use chrono::{DateTime, Utc};
use shared::domain::category::{Category, CategoryId};
use std::collections::HashMap;
use uuid::Uuid;

#[derive(Debug)]
pub struct RawCategory {
    pub id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: String,
    pub index: i16,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub image_count: i64,
    pub jig_count: i64,
}

pub fn build_tree(mut categories: Vec<RawCategory>) -> Vec<Category> {
    let mut parent_to_id_index: HashMap<Option<Uuid>, Vec<usize>> = HashMap::new();

    for (idx, category) in categories.iter().enumerate() {
        parent_to_id_index
            .entry(category.parent_id)
            .and_modify(|it| it.push(idx))
            .or_insert_with(|| vec![idx]);
    }

    parent_to_id_index
        .values_mut()
        .for_each(|it| it.sort_unstable_by_key(|v| categories[*v].index));

    build_tree_recursive(&parent_to_id_index, &mut categories, None)
}

fn build_tree_recursive(
    parent_to_id_index: &HashMap<Option<Uuid>, Vec<usize>>,
    categories: &mut Vec<RawCategory>,
    seed_id: Option<Uuid>,
) -> Vec<Category> {
    let indecies = match parent_to_id_index.get(&seed_id) {
        Some(indecies) => indecies,
        None => return Vec::new(),
    };

    indecies
        .iter()
        .copied()
        .map(|category_index| {
            let children = build_tree_recursive(
                parent_to_id_index,
                categories,
                Some(categories[category_index].id),
            );

            let raw: &mut RawCategory = &mut categories[category_index];

            Category {
                id: CategoryId(raw.id),
                name: raw.name.clone(),
                created_at: raw.created_at,
                updated_at: raw.updated_at,
                image_count: raw.image_count as u64,
                jig_count: raw.jig_count as u64,
                children,
            }
        })
        .collect()
}

#[derive(Debug, Copy, Clone, sqlx::Type)]
#[repr(u16)]
pub enum RegistrationStatus {
    /// The user was just registered ([`POST /v1/user`](shared::api::endpoints::user::Register))
    ///
    /// This state is *skipped* via ouath (they're automatically [`Validated`](Self::Validated))
    /// In this state the user is allowed to:
    /// * Request a new verification email
    /// * Request a password reset(?)
    New = 0,

    /// The user has gone through OAuth and created a new user, or has finished email verification.
    ///
    /// In this state the user is allowed to:
    /// * Request a password reset
    /// * Complete registration via creating their profile
    /// * Delete their account
    Validated = 1,

    /// The user has *completely* finished signup, and their profile has been completed.
    ///
    /// In this state the user is allowed to do anything that their scopes allow.
    /// All users can:
    /// * Change their email (not implemented nor currently on the roadmap)
    /// * Change their password if logged in, reset their password if not.
    /// * Access the API
    /// * Change their profile details (not currently implemented)
    /// * Delete their account
    Complete = 2,
}
