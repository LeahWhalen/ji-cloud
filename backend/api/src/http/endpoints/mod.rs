pub mod admin;
pub mod audio;
pub mod category;
pub mod image;
pub mod jig;
pub mod meta;
pub mod module;
pub mod user;
pub mod animation;
