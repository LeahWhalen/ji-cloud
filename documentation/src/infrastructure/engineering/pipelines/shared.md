# Common Developer Pipeline

## Shared types

Edit the shared Rust types as needed. Request/Response are the primary things that are shared between frontend/backend