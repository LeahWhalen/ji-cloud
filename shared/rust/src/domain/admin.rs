//! Types for admin routes.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::media::{MediaKind, MediaLibrary};

/// Query for [`ListMedia`](crate::api::endpoints::admin::ListMedia)
/// Super unstable, may change at any time, for any reason.
#[derive(Serialize, Deserialize, Debug, Clone, Default)]
#[cfg_attr(feature = "backend", derive(paperclip::actix::Apiv2Schema))]
pub struct AdminListMediaQuery {
    /// Filter to *only* using this specific media kind
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter_kind: Option<MediaKind>,

    /// Filter out anything that was uploaded after this time
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_uploaded_at: Option<DateTime<Utc>>,
}

/// Response for [`ListMedia`](crate::api::endpoints::admin::ListMedia)
/// Super unstable, may change at any time, for any reason.
#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "backend", derive(paperclip::actix::Apiv2Schema))]
pub struct AdminListMediaResponse {
    /// A list of all media items
    pub media: Vec<AdminMediaItem>,
}

/// Item of media
#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "backend", derive(paperclip::actix::Apiv2Schema))]
pub struct AdminMediaItem {
    /// The ID of the media
    pub id: Uuid,

    /// What kind of media this is
    pub kind: MediaKind,

    /// What library this media is from
    pub library: MediaLibrary,

    /// When the media was originally created.
    pub created_at: DateTime<Utc>,

    /// When the media was last updated.
    pub updated_at: Option<DateTime<Utc>>,

    /// When the media was last uploaded.
    pub uploaded_at: Option<DateTime<Utc>>,

    /// An arbitrary (ascii) string representing the current state of the media.
    pub file_etag: Option<String>,
}
