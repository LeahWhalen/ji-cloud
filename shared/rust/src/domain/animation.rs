//! Types for animations.

#[cfg(feature = "backend")]
use paperclip::actix::Apiv2Schema;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Wrapper type around [`Uuid`], represents the ID of an animation.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct AnimationId(pub Uuid);

into_uuid![AnimationId];
