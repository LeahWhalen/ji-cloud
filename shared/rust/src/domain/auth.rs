//! Types for authorization.

use super::meta::{AffiliationId, AgeRangeId, SubjectId};
#[cfg(feature = "backend")]
use paperclip::actix::Apiv2Schema;
use serde::{Deserialize, Serialize};

/// The name to use for JWT cookies.
pub const AUTH_COOKIE_NAME: &str = "X-AUTH";

/// The name of the CSRF header.
pub const CSRF_HEADER_NAME: &str = "X-CSRF";

#[derive(Serialize, Deserialize)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
#[cfg_attr(feature = "backend", openapi(empty))]
/// Request to be sent when registering a new user.
pub struct RegisterRequest {
    /// The user's username.
    ///
    /// This must be unique.
    pub username: String,

    /// Is the user >= 18 yeas old?
    pub over_18: bool,

    /// The user's given name / "first name".
    pub given_name: String,

    /// The user's family name / "last name".
    pub family_name: String,

    // todo: create a struct that enforces format like `en_us`
    /// the language the user prefers to communicate with.
    pub language: String,

    /// The locale that should be used for the user.
    pub locale: String,

    /// the timezone that the user uses.
    pub timezone: chrono_tz::Tz,

    // todo: does this have something to do with emails?
    /// Does the user want educational resources sent to them?
    pub opt_into_edu_resources: bool,

    /// The organization that the user belongs to.
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub organization: Option<String>,

    /// The user's taught subjects.
    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub subjects: Vec<SubjectId>,

    /// The user's age-ranges.
    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub age_ranges: Vec<AgeRangeId>,

    /// The user's affiliations.
    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub affiliations: Vec<AffiliationId>,

    /// The user's location
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub location: Option<serde_json::Value>,
}
