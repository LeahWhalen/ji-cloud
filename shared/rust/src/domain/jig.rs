//! Types for JIGs.

pub mod module;

use super::{meta::ContentTypeId, Publish};
use chrono::{DateTime, Utc};
#[cfg(feature = "backend")]
use paperclip::actix::Apiv2Schema;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// avoid breaking Changes
pub use module::{LiteModule, ModuleId, ModuleKind};

/// Wrapper type around [`Uuid`], represents the ID of a JIG.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct JigId(pub Uuid);

/// Special parameter for allowing implicit `me` as a user.
#[derive(Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
#[serde(rename_all = "camelCase")]
#[serde(untagged)]
pub enum UserOrMe {
    /// We should use the user found in the session auth.
    Me,

    /// we should use the provided user.
    User(Uuid),
}

/// Request to create a new JIG.
#[derive(Serialize, Deserialize, Debug, Default)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct JigCreateRequest {
    /// The JIG's name.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub display_name: Option<String>,

    /// The JIG's remaining modules.
    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(default)]
    pub modules: Vec<ModuleId>,

    /// The types of content this JIG contains.
    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(default)]
    pub content_types: Vec<ContentTypeId>,

    /// When the JIG should be considered published (if at all).
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub publish_at: Option<Publish>,
}

/// The over-the-wire representation of a JIG.
#[derive(Serialize, Deserialize, Clone, Debug)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct Jig {
    /// The ID of the JIG.
    pub id: JigId,

    /// The JIG's name.
    pub display_name: Option<String>,

    /// The JIG's remaining modules.
    pub modules: Vec<LiteModule>,

    /// The types of content this JIG contains.
    pub content_types: Vec<ContentTypeId>,

    /// The ID of the JIG's original creator ([`None`] if unknown).
    pub creator_id: Option<Uuid>,

    /// The current author
    pub author_id: Option<Uuid>,

    /// When the JIG should be considered published (if at all).
    pub publish_at: Option<DateTime<Utc>>,
}

/// The response returned when a request for `GET`ing a jig is successful.
#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct JigResponse {
    /// The requested JIG.
    pub jig: Jig,
}

/// Request for updating a JIG.
#[derive(Serialize, Deserialize, Debug, Default)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct JigUpdateRequest {
    /// The JIG's name.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub display_name: Option<String>,

    /// The JIG's modules.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub modules: Option<Vec<ModuleId>>,

    /// The types of content this JIG contains.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub content_types: Option<Vec<ContentTypeId>>,

    /// The current author
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub author_id: Option<Uuid>,

    /// When the JIG should be considered published (if at all).
    #[serde(deserialize_with = "super::deserialize_optional_field")]
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub publish_at: Option<Option<Publish>>,
}

/// Query for [`Browse`](crate::api::endpoints::jig::Browse).
#[derive(Serialize, Deserialize, Clone, Debug, Default)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
#[serde(rename_all = "camelCase")]
pub struct JigBrowseQuery {
    /// Optionally filter by `is_published`
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_published: Option<bool>,

    /// Optionally filter by author id.
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub author_id: Option<UserOrMe>,

    /// The page number of the jigs to get.
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page: Option<u32>,
}
/// Response for [`Browse`](crate::api::endpoints::jig::Browse).
#[derive(Serialize, Deserialize, Clone, Debug)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
#[serde(rename_all = "camelCase")]
pub struct JigBrowseResponse {
    /// the jigs returned.
    pub jigs: Vec<Jig>,

    /// The number of pages found.
    pub pages: u32,

    /// The total number of jigs found
    pub total_image_count: u64,
}

into_uuid![JigId];
