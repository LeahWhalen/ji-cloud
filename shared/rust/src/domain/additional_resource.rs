//! Types for additional resources for JIGs.

use crate::domain::{
    audio::AudioId, image::ImageId, jig::JigId, learning_path::LearningPathId,
    meta::ResourceTypeId, pdf::PdfId,
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Wrapper type around [`Uuid`](Uuid), represents the ID of an additional resource.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
pub struct AdditionalResourceId(pub Uuid);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
/// Over-the-wire representation of Additional resource.
pub struct AdditionalResource {
    /// The additional resources's ID.
    pub id: AdditionalResourceId,

    /// Name for additional resource
    pub display_name: String,

    /// Type of additional resource
    pub resource_type_id: ResourceTypeId,

    /// Content of additional resource
    #[serde(flatten)]
    pub resource_content: ResourceContent,
}

/// Additional Resources assignment to a JIG or Learning Path
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum JigOrPath {
    /// Additional Resource is assigned to a JIG
    Jig = 0,
    /// Additional Resource is assigned to a Learning Path
    LearningPath = 1,
}

/// Additional Resources assignment to a JIG or Learning Path
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum JigOrPathId {
    /// Additional Resource is assigned to a JIG
    JigId(JigId),
    /// Additional Resource is assigned to a Learning Path
    LearningPathId(LearningPathId),
}
/// Request to create a new `AdditionalResource`.
///
/// [`additional_resource::Create`](crate::api::endpoints::additional_resource::Create)
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceCreateRequest {
    /// Display name for additional resource
    pub display_name: String,

    /// JIG Id or Learning Path Id for additional resource
    #[serde(flatten)]
    pub jig_or_path_id: JigOrPathId,

    /// Type of additional resource
    pub resource_type_id: ResourceTypeId,

    /// Value of additional resource
    #[serde(flatten)]
    pub resource_content: ResourceContent,
}

/// Request to update an `AdditionalResource`.
///
/// [`additional_resource::Update`](crate::api::endpoints::additional_resource::Update)
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceUpdateRequest {
    /// Location of Additional Resource by Id
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(flatten)]
    pub jig_or_path_id: Option<JigOrPathId>,

    /// resource display name
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub display_name: Option<String>,

    /// Type of additional  resource
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub resource_type_id: Option<ResourceTypeId>,

    /// Kind of additional resource
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(flatten)]
    pub resource_content: Option<ResourceContent>,
}

/// Request to get `AdditionalResource` associated with a live Jig or Learning Path.
///
/// [`additional_resource::Update`](crate::api::endpoints::additional_resource::GetLive)
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceGetLiveQuery {
    /// Location of Additional Resource by Id
    pub jig_or_path: JigOrPath,

    /// Jig Id Or Learning Path Id
    pub id: Uuid,
}

/// Request to get `AdditionalResource` associated with a draft Jig or Learning Path.
///
/// [`additional_resource::AdditionalResourceGetDraftQuery`](crate::api::endpoints::additional_resource::GetDraft)
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceGetDraftQuery {
    /// Location of Additional Resource by Id
    pub jig_or_path: JigOrPath,

    /// JIG id Or Learning Path id
    pub id: Uuid,
}

/// Request to delete an `AdditionalResource`.
///
/// [`additional_resource::DeleteAdditionalResourceRequest`](crate::api::endpoints::additional_resource::Delete)
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct DeleteAdditionalResourceRequest {
    /// Location of Additional Resource by Id
    pub jig_or_path: JigOrPath,

    /// JIG id Or Learning Path id
    pub id: Uuid,
}

/// Value of additional resource
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ResourceContent {
    /// Additional resource kind: image
    ImageId(ImageId),
    /// Additional resource kind: audioFile
    AudioId(AudioId),
    /// Additional resource kind: link
    Link(url::Url),
    /// Additional resource kind: pdf
    PdfId(PdfId),
}

into_uuid![AdditionalResourceId];
