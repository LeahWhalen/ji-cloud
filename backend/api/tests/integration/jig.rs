use http::StatusCode;
use shared::domain::{jig::JigId, CreateResponse};

use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
};

#[actix_rt::test]
async fn create_default() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User]).await;

    let port = app.port();

    let _ = tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: CreateResponse<JigId> = resp.json().await?;

    insta::assert_json_snapshot!(body, {".id" => "[id]"});

    Ok(())
}

// todo: create_params
// todo: delete
// todo: get
// todo: update
