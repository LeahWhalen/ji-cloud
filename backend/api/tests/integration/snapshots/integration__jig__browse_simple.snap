---
source: tests/integration/jig.rs
expression: body

---
{
  "jigs": [
    {
      "id": "0cc084bc-7c83-11eb-9f77-e3218dffb008",
      "display_name": null,
      "modules": [
        {
          "id": "0cbfdd82-7c83-11eb-9f77-d7d86264c3bc",
          "kind": "Cover"
        },
        {
          "id": "0cc03a02-7c83-11eb-9f77-f77f9ad65e9a",
          "kind": null
        }
      ],
      "goals": [],
      "creator_id": "1f241e1b-b537-493f-a230-075cb16315be",
      "author_id": "1f241e1b-b537-493f-a230-075cb16315be",
      "language": "en",
      "publish_at": null
    }
  ],
  "pages": 1,
  "totalJigCount": 1
}
